thin-replica-server/include/thin-replica-server/thin_replica_impl.hpp

```diff
@@ -933,72 +933,84 @@ class ThinReplicaImpl {
+        auto filtered_eg_update = kvb_filter->filterEventGroupUpdate(live_updates->oldestEventGroup());
+        // If the oldest live update is not for the requesting client, keep reading from the live update queue
+        // until the first relevant live update is reached. Drop all non-relevant live updates read along the way.
+        if (!filtered_eg_update) {
           SubEventGroupUpdate sub_eg_update;
           live_updates->PopEventGroup(sub_eg_update);
+          LOG_INFO(logger_, "Sync dropping upon filtering " << sub_eg_update.event_group_id);
+          num_updates_filtered_out++;
           is_update_available = false;
+          next_global_eg_id_to_read += 1;
+          continue;
         }
+        break;
       }
-    }
-    ConcordAssertGE(newest_eg_id, end + 1);
-    ConcordAssertGT(live_updates->oldestEventGroupId(), last_global_eg_id_read);
-    auto global_eg_id_to_read = kvb_filter->findGlobalEventGroupId(end + 1).global_id;
-    while (live_updates->oldestEventGroupId() < global_eg_id_to_read) {
-      // Drop updates from the live-update-queue if the oldest global-eg-id on the live-update-queue is less than the
-      // next global-eg-id to be read (corresponds to external-eg-id `end+1`)
-      SubEventGroupUpdate sub_eg_update;
-      live_updates->PopEventGroup(sub_eg_update);
-      if (live_updates->EmptyEventGroupQueue()) {
-        is_update_available = live_updates->waitForEventGroupUntilNonEmpty(kWaitForUpdateTimeout);
-      }
-    }
-    // We are in sync already if the oldest event global-eg-id on live_updates is same as next global_eg_id_to_read
-    if (live_updates->oldestEventGroupId() == global_eg_id_to_read) {
-      ConcordAssertGE(global_eg_id_to_read, live_updates->oldestEventGroupId());
-      ConcordAssertLE(global_eg_id_to_read, live_updates->newestEventGroupId());
-      return;
-    }
 
-    // Gap:
-    // If the first live update is not the follow-up to the last read event group from
-    // KVB then we need to fill the gap. Let's read from KVB starting at end + 1
-    // up to updates that are part of the live updates already. Thereby, we
-    // create an overlap between what we read from KVB and what is currently in
-    // the live updates.
-    if (live_updates->oldestEventGroupId() > global_eg_id_to_read) {
-      ConcordAssertGT(kvb_filter->newestExternalEventGroupId(), end + 1);
-      start = end + 1;
-      end = kvb_filter->newestExternalEventGroupId();
+      // fill the gap
+      LOG_INFO(logger_,
+               "Sync filling gap [" << next_global_eg_id_to_read << ", " << live_updates->oldestEventGroupId() << "]");
+      while (live_updates->oldestEventGroupId() > next_global_eg_id_to_read) {
+        // gap exists between next_global_eg_id_to_read and live_updates->oldestEventGroupId()
+        // Let's read from data table directly, filter and send updates, until next_global_eg_id_to_read equals
+        // live_updates->oldestEventGroupId()
```

Comment:
I suggest to put this comment with the "fill the gap" comment or more specifically, above the while loop.